sgkit/variables.py

```diff
@@ -0,0 +1,225 @@
+from dataclasses import dataclass
+from typing import Dict, Hashable, Mapping, Set, Union, overload
+
+import xarray as xr
+
+
+@dataclass(frozen=True)
+class Spec:
+    """Root type Spec"""
+
+    default_name: str
+
+
+@dataclass(frozen=True)
+class ArrayLikeSpec(Spec):
+    """ArrayLike type spec"""
+
+    kind: Union[None, str, Set[str]] = None
+    ndim: Union[None, int, Set[int]] = None
+
+
+call_genotype = ArrayLikeSpec("call_genotype", kind="i", ndim=3)
+"""
+Call genotype. Encoded as allele values (0 for the reference, 1 for
+the first allele, 2 for the second allele), or -1 to indicate a
+missing value.
+"""
+call_genotype_mask = ArrayLikeSpec("call_genotype_mask", kind="b", ndim=3)
+"""TODO"""
+variant_contig = ArrayLikeSpec("variant_contig", kind="i", ndim=1)
+"""The (index of the) contig for each variant."""
+variant_position = ArrayLikeSpec("variant_position", kind="i", ndim=1)
+"""The reference position of the variant."""
+variant_allele = ArrayLikeSpec("variant_allele", kind={"S", "O"}, ndim=2)
+"""The possible alleles for the variant."""
+sample_id = ArrayLikeSpec("sample_id", kind={"U", "O"}, ndim=1)
+"""The unique identifier of the sample."""
+call_genotype_phased = ArrayLikeSpec("call_genotype_phased", kind="b", ndim=2)
+"""
+A flag for each call indicating if it is phased or not. If omitted
+all calls are unphased.
+"""
+variant_id = ArrayLikeSpec("variant_id", kind="U", ndim=1)
+"""The unique identifier of the variant."""
+call_dosage = ArrayLikeSpec("call_dosage", kind="f", ndim=2)
+"""Dosages, encoded as floats, with NaN indicating a missing value."""
+call_dosage_mask = ArrayLikeSpec("call_dosage_mask", kind="b", ndim=2)
+"""TODO"""
+call_genotype_probability = ArrayLikeSpec("call_genotype_probability", kind="f", ndim=3)
+"""TODO"""
+call_genotype_probability_mask = ArrayLikeSpec(
+    "call_genotype_probability_mask", kind="b", ndim=3
+)
+"""TODO"""
+genotype_counts = ArrayLikeSpec("genotype_counts", ndim=2, kind="i")
+"""
+Genotype counts. Must correspond to an (`N`, 3) array where `N` is equal
+to the number of variants and the 3 columns contain heterozygous,
+homozygous reference, and homozygous alternate counts (in that order)
+across all samples for a variant.
+"""
+call_allele_count = ArrayLikeSpec("call_allele_count", ndim=3, kind="u")
+"""
+Allele counts. With shape (variants, samples, alleles) and values
+corresponding to the number of non-missing occurrences of each allele.
+"""
+variant_allele_count = ArrayLikeSpec("variant_allele_count", ndim=2, kind="u")
+"""
+Variant allele counts. With shape (variants, alleles) and values
+corresponding to the number of non-missing occurrences of each allele.
+"""
+variant_hwe_p_value = ArrayLikeSpec("variant_hwe_p_value", kind="f")
+"""P values from HWE test for each variant as float in [0, 1]."""
+variant_beta = ArrayLikeSpec("variant_beta")
```

Comment:
For this and the other "naming" comments below + all the TODOs in the docs, I suggest we open a separate issue and follow up in separate PR(s). One of the nice side effects of having all those variables in the same place is that we can see the inconsistencies etc. wdyt @tomwhite ?

Reply:
Yes, fine by me

Reply:
@tomwhite so I just remembered something, this would be going against some early comments in https://github.com/pystatgen/sgkit/issues/17, what do you think? also ping @eric-czech

Reply:
@ravwojdyla do you mean the comment about `variables.pc_relate_phi.default_name` vs `"pc_relate_phi"`? OK, happy to leave as a string.

Reply:
@tomwhite right, wrong comment section, sorry, this should have been in https://github.com/pystatgen/sgkit/pull/276#discussion_r495782207

Let's see what @eric-czech thinks. Other than that, afaiu @tomwhite this is ready for another review.